pkg/cmd/release-payload-controller/prowjob_controller.go

```diff
@@ -0,0 +1,210 @@
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return
+	}
+
+	klog.Infof("Processing prowjob: '%s/%s'", namespace, name)
+
+	// Get the ProwJob resource with this namespace/name
+	prowJob, err := c.prowJobLister.ProwJobs(namespace).Get(name)
+	if err != nil {
+		// The ProwJob resource may no longer exist, in which case we stop processing.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("prowJob '%s' no longer exists", key))
+			return
+		}
+		utilruntime.HandleError(fmt.Errorf("unable to retreive prowJob '%s': %v", key, err))
+		return
+	}
+
+	tag, ok := prowJob.Annotations["release.openshift.io/tag"]
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unable to process prowjob, missing 'release.openshift.io/tag' annotation"))
+		return
+	}
+
+	releasePayloadKey := fmt.Sprintf("%s/%s", c.releasePayloadNamespace, tag)
+	klog.Infof("Queueing ReleasePayload: %s", releasePayloadKey)
+	c.queue.Add(releasePayloadKey)
+}
+
+func (c *ProwJobStatusController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	klog.Info("Starting ProwJob Status Controller")
+	defer func() {
+		klog.Info("Shutting down ProwJob Status Controller")
+		c.queue.ShutDown()
+		klog.Info("ProwJob Status Controller shut down")
+	}()
+
+	if !cache.WaitForNamedCacheSync("ProwJobStatusController", ctx.Done(), c.cachesToSync...) {
+		return
+	}
+
+	go func() {
+		wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}()
+
+	<-ctx.Done()
+}
+
+func (c *ProwJobStatusController) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *ProwJobStatusController) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.sync(ctx, key.(string))
+
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %w", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *ProwJobStatusController) sync(ctx context.Context, key string) error {
+	klog.V(4).Infof("Starting ProwJobStatusController sync")
+	defer klog.V(4).Infof("ProwJobStatusController sync done")
+
+	// Always start 10 seconds later after a change occurred. Makes us less likely to steal work and logs from the operator.
+	timer := time.NewTimer(10 * time.Second)
+	defer timer.Stop()
+	select {
+	case <-timer.C:
+	case <-ctx.Done():
+		return nil
+	}
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	klog.Infof("Processing ReleasePayload: '%s/%s' from workQueue", namespace, name)
+
+	// Get the ReleasePayload resource with this namespace/name
+	releasePayload, err := c.releasePayloadLister.ReleasePayloads(namespace).Get(name)
```

Comment:
suggest naming `originalReleasePayload` and a few lines down you do `releasePayload = originalReleasePay.oad.DeepCopy`